### ISSUE TITLE
5360 - Unique - the shortest

### ISSUE THREAD

**Issue by Alexsey** [type-challenges/type-challenges#21766]
Iterates over `T` by one item
`U` - the union of already included items wrapped into arrays. E.g. for `T` equals `['a', string, ...]`, on step 2 the `U` would be `['a'] | [string]`. We need to wrap items into arrays because otherwise `'a' | string` would collapse into `string`

```ts
type Unique<T, U = never> =
  T extends [infer F, ...infer R]
    ? true extends (U extends U ? Equal<U, [F]> : never)
      ? Unique<R, U>
      : [F, ...Unique<R, U | [F]>]
    : []
```

**Comment by likepeach789** [comment]
Could you help explain this line `true extends (1 extends 1 ? Equal1<U, [F]> : never)` , thanks.

**Comment by Alexsey** [comment]
@likepeach789

**As for your code**:
The reason why your type resolves to `never` is because during the first recursion call:
* `U` is `never`
* `U extends U ? ... : never` is `never extends never ? ... : never` is `never`
* `true extends never ? 1111 : never` is `never`

So as you can see, the first recursion call always goes to the `[F, ...Unique<R, U | [F]>` branch, and this is how `U` gets its first element `[F]` that is the first item of `T` array

**How the line works** `true extends (U extends U ? Equal1<U, [F]> : never)`:

* `U extends U` is a common pattern to distribute the union type `U`, so the `U` inside of the "truthy" branch is not `U` but an individual union element
* `U` starts as an empty union (`never`) and populates **only** in line `[F, ...Unique<R, U | [F]>]` where `F` are items of the input array `T` wrapped in arrays


Let's take an example:
`T` equals `['a', string, 'a']`
The first recursion call:
* `U` is `never`, `F` is `'a'`, `R` is `[string, 'a']`
* Because `U` is `never`, the outer conditional type resolves to `[F, ...Unique<R, U | [F]>]` branch, that is `['a', ...Unique<[string, 'a'], never | ['a']]>`
* 
The second recursion call:
* `U` is `['a']`, `F` is `string`, `R` is `['a']`
* The outer conditional type goes like `true extends Equal<['a'], string> ? ... : ...` so it resolves to `[F, ...Unique<R, U | [F]>]` branch, that is `[string, ...Unique<['a'], ['a'] | string]>`
* 
The third recursion call:
* `U` is `['a'] | [string]`, `F` is `'a'`, `R` is `[]`
* Distributed outer conditional type goes like `true extends (Equal<['a'], ['a']> | Equal<[string], ['a']>) ? ... : ...` that is `true extends (true | false) ? ... : ...` that resolves to `Unique<R, U>` branch, that is `Unique<[], ['a'] | string>`
* 
The fourth recursion call:
* * `U` is `['a'] | [string]`, `T` is `never`, so resolves to `[]`

**Comment by likepeach789** [comment]
@Alexsey Thank you SO MUCH for the detailed answer! The explanation was clear and thorough.  Just to make sure i understand it correctly, below code snippet should make sense, right?
```
type A<T, U> = true extends (T extends T ? Equal<T, U> : never) ? 1 : 2
type B = A<[1] | [2], [2]>; // 1

// Analyze ----------------
type B = true extends (([1] extends [1] | [2] ? Equal<[1], [2]> : never) | ([2] extends [1] | [2] ? Equal<[2], [2]> : never)) ? 1 : 2
       = true extends ((Equal<[1], [2]>) | (Equal<[2], [2]>)) ? 1 : 2
       = true extends (false | true) ? 1 : 2
       = 1
```

**Comment by Alexsey** [comment]
@likepeach789 Yes, I would expect it to resolve in a way close to the one you wrote. The old handbook has an [exhaustive section on this topic](https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types)

**Comment by likepeach789** [comment]
That's great! Thank you Alexsey. ☺

**Comment by goddnsgit** [comment]
你的代码太漂亮了。

**Comment by GordanX** [comment]
The force solution,hh

```ts
type IsEqual<T, U> = (<G>() => G extends T ? true : false) extends (<G>() => G extends U ? true : false) ? true : false
type Contains<T extends any[], U> = T extends [...infer R, infer F] ? IsEqual<F, U> extends true ? true : Contains<R, U> : false
type Unique<T extends any[]> = T extends [...infer R, infer L] ? Contains<R, L> extends true ? Unique<R> : [...Unique<R>, L] : []
```
